### PR TITLE
intake: Add `context.message.routing_key` to API

### LINKED ISSUE #5229: Updating messaging context fields according to spec
As per our [messaging spec](https://github.com/elastic/apm/blob/master/specs/agents/tracing-instrumentation-messaging.md), we should have a `context.message.routing_key`  field both for transactions and spans (optional, string). 

In addition, `context.message.body` and `context.message.headers` exist in both spans and transactions, but they are only relevant to transactions (similar to the corresponding fields in `context.http`).

### PR DESCRIPTION
**Pull request by marclop**
## Motivation/summary

Adds `{span,transaction}.context.message.routing_key` as two optional
string fields to the Intake API.

## Checklist

- [x] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/master/CHANGELOG.asciidoc)
- [x] Documentation has been updated

### Setup

1. `make && docker compose up -d`
3. Run APM server: `./apm-server -E output.elasticsearch.username=admin -E output.elasticsearch.password=changeme -e`
4. Ingest both `testdata/intake-v2/transactions.ndjson` and `testdata/intake-v2/spans.ndjson`:
    - `curl -H "Content-type: application/x-ndjson" --data-binary @testdata/intake-v2/transactions.ndjson http://localhost:8200/intake/v2/events`
    - `curl -H "Content-type: application/x-ndjson" --data-binary @testdata/intake-v2/spans.ndjson http://localhost:8200/intake/v2/events`

### Run the following query in `Dev Tools > Console`
```
GET /apm*-span,apm*-transaction/_search
{
  "_source": [false], 
  "fields": [
    "transaction.message.routing_key",
    "span.message.routing_key"
  ], 
  "query": {
    "bool": {
      "should": [
        {
          "match_phrase": {
            "transaction.type": "messaging"
          }
        },
        {
          "match_phrase": {
            "span.type": "messaging"
          }
        }
      ]
    }
  }
}
```

Assert that there are 2 results and that under `"fields"`:

-  `"span.messagge.routing_key" == "user-created-span"`
-  `"transaction.messagge.routing_key" == "user-created-transaction"`

## Related issues

Closes #5229

### CODE REVIEW

**Comment by marclop** [comment]
I was unable to test this locally using the existing `docker-compose.yml` in the root of the repository, I suspect it has something to do with my M1 MacBook. The `apm-server_package-registry_1` container segfaulted but I didn't have time to debug it:

```
$ docker logs apm-server_package-registry_1
...snip...
2021/09/15 16:59:49 zscaler             	     0.3.2	/packages/snapshot/zscaler/0.3.2
2021/09/15 16:59:49 zscaler             	     0.3.3	/packages/snapshot/zscaler/0.3.3
2021/09/15 16:59:49 Packages in /packages/local:
2021/09/15 16:59:49 apm                 	     0.5.0	/packages/local/apm/0.5.0
qemu: uncaught target signal 11 (Segmentation fault) - core dumped
```

Tomorrow I'll try to test this change using the `apm-integration-testing` repo.

File: model/modeldecoder/v2/model.go [hunk 1]
```diff
@@ -106,6 +106,9 @@ type contextMessage struct {
 	Headers nullable.HTTPHeader `json:"headers"`
 	// Queue holds information about the message queue where the message is received.
 	Queue contextMessageQueue `json:"queue"`
+	// RoutingKey holds the optional routing key of the received message as set
+	// on the queuing system. Currently only used on for RabbitMQ messages.
```
Comment (marclop) [inline model/modeldecoder/v2/model.go:110]: I'm not entirely convinced about the last sentence about RabbitMQ. Removing it might be better than tying this to a specific queuing technology. It's only referenced in the description of the field, but still...
```suggestion
	// RoutingKey holds the optional routing key of the received message as set
	// on the queuing system.
```

File: model/modeldecoder/v2/decoder.go [hunk 2]
```diff
@@ -1011,6 +1014,10 @@ func mapToTransactionModel(from *transaction, event *model.APMEvent) {
 			if from.Context.Message.Queue.IsSet() && from.Context.Message.Queue.Name.IsSet() {
 				out.Message.QueueName = from.Context.Message.Queue.Name.Val
 			}
+			if from.Context.Message.RoutingKey.IsSet() {
+				println(from.Context.Message.RoutingKey.Val)
```
Comment (axw) [inline model/modeldecoder/v2/decoder.go:1018]:
```suggestion
```

printf debugging crew :100:

File: model/modeldecoder/v2/model.go (hunk 1 above)
Comment (axw) [reply model/modeldecoder/v2/model.go:110]: Maybe reword it as an example? e.g.

> // RoutingKey holds the optional routing key of the received message as set
> // on the queuing system, such as in RabbitMQ.

**Review by axw — COMMENTED** [review]
> The requested field in #5229 is only for the transactions and not the spans

That issue says "... we should have a context.message.routing_key field both for transactions and spans ..."

I think your change looks good.

File: model/modeldecoder/v2/decoder.go (hunk 2 above)
Comment (marclop) [reply model/modeldecoder/v2/decoder.go:1018]: ouch, will remove

**Comment by marclop** [comment]
@axw I mixed up the requested new field and two of the fields that currently exist within the `message` object, I was referring to what Eyal mentions:

> In addition, `context.message.body` and `context.message.headers` exist in both spans and transactions, but they are only relevant to transactions (similar to the corresponding fields in `context.http`).

**Comment by marclop** [comment]
Added back the FaaS changelog entry that seems to have been removed in https://github.com/elastic/apm-server/commit/93efb5cc2662f80138beaae8faf009ddc60ef252

**Review by axw — APPROVED** [review]
looks great!

**Review by axw — APPROVED** [review]
